doc/designs/quic-design/quic-api.md

```diff
@@ -818,6 +818,219 @@ reason string in bytes.
+ */
+int SSL_is_connection(SSL *ssl);
+```
+
+#### `SSL_get_stream_type`
+
+```c
+/*
+ * If the object represents a stream, returns a SSL_STREAM_TYPE value
+ * designating whether the stream can be used for transmission, reception,
+ * or both.
+ *
+ * This always returns SSL_STREAM_TYPE_BIDI for non-QUIC methods.
+ *
+ * It returns SSL_STREAM_TYPE_NONE for a QUIC connection object if it
+ * does not have a default stream.
+ */
+#define SSL_STREAM_TYPE_NONE    0
+#define SSL_STREAM_TYPE_READ    1
+#define SSL_STREAM_TYPE_WRITE   2
+#define SSL_STREAM_TYPE_BIDI    (SSL_STREAM_TYPE_READ | SSL_STREAM_TYPE_WRITE)
+__owur int SSL_get_stream_type(SSL *ssl);
+```
+
+#### `SSL_get_stream_id`
+
+```c
+/*
+ * QUIC: Returns the unique stream ID for the stream, an integer in range [0,
+ * 2**62-1]. If called on a QUIC connection, returns the unique stream ID for
+ * the default stream if there is one, and otherwise returns 0.
+ *
+ * TLS, DTLS: Returns 0.
+ */
+__owur uint64_t SSL_get_stream_id(SSL *ssl);
+```
+
+#### `SSL_new_stream`
+
+```c
+/*
+ * Create a new SSL object representing a single additional stream.
+ *
+ * There is no need to call SSL_connect on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Creates a new stream. Must be called only on a QUIC connection SSL object.
+ *   Can be used on client or server. If the SSL_FLAG_UNI flag is set, the
+ *   created stream is unidirectional, otherwise it is bidirectional.
+ *
+ *   To be MSMT-safe.
+ *
+ * For TLS and DTLS SSL objects:
+ *   Always fails.
+ */
+#define SSL_FLAG_UNI    1
+
+SSL *SSL_new_stream(SSL *ssl, uint64_t flags);
+```
+
+#### `SSL_accept_stream`
+
+```c
+/*
+ * Create a new SSL object representing an additional stream which has created
+ * by the peer.
+ *
+ * There is no need to call SSL_accept on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Must be called only on a QUIC connection SSL object. Fails if called on a
+ *   stream object. Checks if a new stream has been created by the peer. If it
+ *   has, creates a new SSL object to represent it and returns it. Otherwise,
+ *   returns NULL.
+ *
+ * For all other methods:
+ *   Returns NULL.
+ *
+ * The flags argument is unused and should be set to zero.
+ *
+ * To be MSMT-safe (i.e., can be called from multiple threads).
+ *
+ * If the QUIC connection SSL object is configured in blocking mode, this
+ * function will block unless the SSL_ACCEPT_STREAM_NO_BLOCK flag is passed.
+ */
+#define SSL_ACCEPT_STREAM_NO_BLOCK      1
+
+SSL *SSL_accept_stream(SSL *ssl, uint64_t flags);
+```
+
+#### `SSL_get_accept_stream_queue_len`
+
+```c
+/*
+ * Determine the number of streams waiting to be returned on a subsequent call
+ * to SSL_accept_stream. If this returns a non-zero value, SSL_accept_stream is
+ * guaranteed to work. Returns 0 for non-QUIC objects, or for QUIC stream
+ * objects.
+ *
+ * To be MSMT-safe.
```

Comment:
There will be a race when this returns a non-zero value that kind of invalidates the guarantee.

I.e. SSL_get_accept_stream_queue_len() returns 1 but SSL_accept_stream() does fail because another thread grabbed it.

Reply:
Yeah, that's valid. You could use `SSL_accept_stream` with `SSL_ACCEPT_STREAM_NO_BLOCK`.